source/particle/generator/ascii_file.cc

```diff
@@ -0,0 +1,173 @@
+  {
+    namespace Generator
+    {
+      template <int dim>
+      AsciiFile<dim>::AsciiFile() {}
+
+      template <int dim>
+      std::multimap<types::LevelInd, Particle<dim> >
+      AsciiFile<dim>::generate_particles()
+      {
+        const std::string filename = data_directory+data_filename;
+        std::ifstream in(filename.c_str(), std::ios::in);
+        AssertThrow (in,
+                     ExcMessage (std::string("Couldn't open data file <"
+                                             +
+                                             filename
+                                             +
+                                             ">.")));
+
+        // Read header lines
+        while (in.peek() == '#')
+          {
+            std::string temp;
+            getline(in,temp);
+          }
+
+        // Read data lines
+        types::particle_index id = 0;
+        Point<dim> coordinates;
+        std::multimap<types::LevelInd, Particle<dim> > particles;
+
+        if (dim == 2)
+          while (in >> coordinates[0] >> coordinates[1])
+            {
+              particles.insert(this->generate_particle(coordinates,id));
+              id++;
+            }
+        else if (dim == 3)
+          while (in >> coordinates[0] >> coordinates[1] >> coordinates[2])
+            {
+              particles.insert(this->generate_particle(coordinates,id));
+              id++;
+            }
+        else
+          Assert(false,ExcNotImplemented());
+
+        return particles;
+      }
+
+
+      template <int dim>
+      void
+      AsciiFile<dim>::declare_parameters (ParameterHandler &prm)
+      {
+        prm.enter_subsection("Postprocess");
+        {
+          prm.enter_subsection("Tracers");
+          {
+            prm.enter_subsection("Generator");
+            {
+              prm.enter_subsection("Ascii file");
+              {
+                prm.declare_entry ("Data directory",
+                                   "$ASPECT_SOURCE_DIR/data/particle/generator/ascii/",
+                                   Patterns::DirectoryName (),
+                                   "The name of a directory that contains the tracer data. This path "
+                                   "may either be absolute (if starting with a '/') or relative to "
+                                   "the current directory. The path may also include the special "
+                                   "text '$ASPECT_SOURCE_DIR' which will be interpreted as the path "
+                                   "in which the ASPECT source files were located when ASPECT was "
+                                   "compiled. This interpretation allows, for example, to reference "
+                                   "files located in the 'data/' subdirectory of ASPECT. ");
+                prm.declare_entry ("Data file name", "tracer.dat",
+                                   Patterns::Anything (),
+                                   "The name of the tracer file.");
```

Comment:
what's the reason for two parameters specifying directory and file? Couldn't you just provide a single parameter that states the entire path?